app/position/providers/androidpositionprovider.cpp

```diff
@@ -0,0 +1,277 @@
+    // may not even show up and we get denied response again.
+    __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] going to request permissions" );
+    qApp->requestPermission( perm, [this]( const QPermission & p )
+    {
+      if ( p.status() == Qt::PermissionStatus::Granted )
+      {
+        __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] permissions granted!" );
+        this->startUpdates();
+      }
+      else
+      {
+        __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] permissions denied :-(" );
+        // User may have granted permission just for approximate location,
+        // so it would be good to detect that and warn about that: approximate
+        // location has intentionally only ~2km accuracy - too low for any data collection
+        QLocationPermission permApprox;
+        permApprox.setAccuracy( QLocationPermission::Approximate );
+        if ( qApp->checkPermission( permApprox ) == Qt::PermissionStatus::Granted )
+          this->setState( tr( "Approximate location only!" ), State::NoConnection );
+        else
+          this->setState( tr( "No location permissions" ), State::NoConnection );
+      }
+    } );
+    return;
+  }
+
+  // TODO: this should not be needed?
+  AndroidPositionProvider::startUpdates();
+}
+
+AndroidPositionProvider::~AndroidPositionProvider()
+{
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "DESTRUCT" );
+
+  Q_ASSERT( sInstances[mInstanceId] == this );
+  sInstances.remove( mInstanceId );
+
+}
+
+bool AndroidPositionProvider::isFusedAvailable()
+{
+  jobject context = QNativeInterface::QAndroidApplication::context();
+
+  return QJniObject::callStaticMethod<jboolean>( "uk/co/lutraconsulting/MMAndroidPosition", "isFusedLocationProviderAvailable",
+         "(Landroid/content/Context;)Z", context );
+}
+
+QString AndroidPositionProvider::fusedErrorString()
+{
+  jobject context = QNativeInterface::QAndroidApplication::context();
+
+  QJniObject str = QJniObject::callStaticObjectMethod( "uk/co/lutraconsulting/MMAndroidPosition", "fusedLocationProviderErrorString",
+                   "(Landroid/content/Context;)Ljava/lang/String;", context );
+
+  return str.toString();
+}
+
+
+void AndroidPositionProvider::startUpdates()
+{
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] start updates" );
+
+  jboolean res = mAndroidPos.callMethod<jboolean>( "start", "()Z" );
+
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] start updates res: %d", res );
+
+  if ( !res )
+  {
+    QJniObject errMsgJni = mAndroidPos.callObjectMethod( "errorMessage", "()Ljava/lang/String;" );
+    QString errMsg = errMsgJni.toString();
+    __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] error: %s", errMsg.toUtf8().constData() );
+    if ( errMsg == "MISSING_PERMISSIONS" )
+      setState( tr( "No location permissions" ), State::NoConnection );
+    else if ( errMsg == "FUSED_NOT_AVAILABLE" )
+      setState( tr( "Fused location not available" ), State::NoConnection );
```

Comment:
Shall we have these strings `MISSING_PERMISSIONS` and `FUSED_NOT_AVAILABLE` defined as `const strings`? Alternatively they could be in the `AbstractPositionProvider`